FirstPersonControls: Added .lookAt()

Mugen87 opened the pull request:
Fixed #15629

looeee commented:
Wow, this is literally the exact thing that's been bugging me for the last half hour. Perfect timing 😀

Mugen87 commented:
Wonderful! 👍

WestLangley commented on examples/js/controls/FirstPersonControls.js:
```diff
@@ -30,6 +30,7 @@ THREE.FirstPersonControls = function ( object, domElement ) {
 	this.verticalMax = Math.PI;
 
 	this.autoSpeedFactor = 0.0;
+	this.forward = new THREE.Vector3( 0, 0, - 1 );
```
What happens if the user changes this?

Mugen87 replied:
Then the forward vector is differently. This is necessary if the user passes a ordinary `Object3D` to the controls.

makc replied:
if used to control other object, it becomes ThirdPersonControls :)

looeee replied:
> I believe FirstPersonControls is designed to control a camera

Why put a limit on what it can control, if it's as simple as exposing `this.forward`?

looeee replied:
If we do decide to impose this limit then we should change this line:

``` js 
THREE.FirstPersonControls = function ( object, domElement ) 
```
to
``` js
THREE.FirstPersonControls = function ( camera, domElement ) 
```

And probably also for other controls too, because `object` suggests that any `Object3D` can be used.

Mugen87 replied:
@looeee Absolutely!

makc replied:
@looeee @Mugen87 I think the proper question is why the hell forward needs to be different for camera vs other objects? Why can't it just be 0,0,-1 for everything?

Mugen87 replied:
That's more or less a WebGL standard. (0,0,1) for ordinary objects and (0,0,-1) for cameras, see https://stackoverflow.com/questions/33407209/why-does-the-camera-face-the-negative-end-of-the-z-axis-by-default?rq=1

Mugen87 replied:
@looeee I have hardwired the forward vector for now since the controls do not fully support arbitrary objects yet, see https://jsfiddle.net/q50o1r97/1/. If there is any interest in changing this, I suggest to do this with a separate PR. Same for renaming `object` to `camera` in control classes.

looeee replied:
@Mugen87 OK, but I think it would be good to have clarity on this. Either we do support controlling arbitrary objects with controls, or we don't. 

If we do, then it should be noted somewhere - even just as a comment in the files. 

If we don't, then we can just rename `object` to `camera`. 
I'm OK with either.

WestLangley commented:
Sorry, I do not see how this can work. It is also not consistent with the three.js API.

Are you saying you can't call `setOrientation()` without calling `camera.lookAt()` first?

Is `camera.lookAt()` and `setOrientation()` supposed to be called only once prior to initializing the controls?

I hope that is the intention -- users should not be setting the camera behind the back of the controls.

The proper solution is to make `setOrientation()` private, and to set the camera orientation in the controls constructor.

Mugen87 commented:
> Are you saying you can't call setOrientation() without calling camera.lookAt() first?

No, it's just an example.

> Is camera.lookAt() and setOrientation() supposed to be called only once prior to initializing the controls?

You can also call if after initializing the controls.

> Sorry, I do not see how this can work. It is also not consistent with the three.js API.

It worked perfectly on my computer. I think the stackoverflow post show that there is a demand for action. So your attitude seems a bit counterproductive to me.

makc commented:
@WestLangley they suggested the method rather than inline code because someone might want to change the camera direction after the controls were activated.

looeee commented:
I've had a chance to test this. It's an improvement IMO, but what I would prefer (and what I believe the other controls do) is to respect the camera's orientation and position at the time when the controls are instantiated. 

If I do: 

```js
camera.lookAt( someObject.position );

const controls = new FirstPersonControls( camera );
```

The current behaviour is that the camera jumps to a new and unexpected orientation. 
With this change, I can now do `controls.setOrientation( camera.quaternion )` to fix this, which is an improvement. But I'd rather that just happens automatically, which again, I think is the case for other controls.

Mugen87 commented:
`FirstPersonControls.lookAt()` is a good idea! I've changed to PR accordingly and also ensured that @looeee suggestion is implemented. This works because `setOrientation()` is now a private function and called when the controls are instantiated.

Mugen87 commented:
`FirstPersonControls.lookAt()` sounds good. I have changed the PR and also implemented @looeee's suggestion. `setOrientation()` is now a private function and called in the ctor. This means it's now possible to respect the original position of the camera.

Mugen87 commented:
I'll make a note in the migration guide about the changes related to `FirstPersonControls`.

Mugen87 commented:
BTW: I have noticed that `FirstPersonControls.constrainVertical` has somewhat a confusing effect when using `.lookAt()`. It alters the value of `phi` and thus the result of `.lookAt()`. I mean it produces the expected result but users should aware about this fact when turning `constrainVertical` to `true`.

WestLangley commented on examples/webgl_nearestneighbour.html:
```diff
@@ -90,6 +90,9 @@
 				controls.movementSpeed = 100;
 				controls.lookSpeed = 0.1;
 
+				controls.lookAt( new THREE.Vector3( 500, 500, 500 ) );
```
For consistency, I think this API should be supported, too:
```javascript
controls.lookAt( 500, 500, 500 );
```

WestLangley commented on examples/js/controls/FirstPersonControls.js:
```diff
@@ -295,6 +313,20 @@ THREE.FirstPersonControls = function ( object, domElement ) {
+
+		var quaternion = controls.object.quaternion;
+
+		lookDirection.copy( controls.forward ).applyQuaternion( quaternion );
```
As I said previously, 
```javascript
lookDirection.set( 0, 0, - 1 ).applyQuaternion( quaternion );
```
This control only supports cameras.

Mugen87 replied:
I have tested it and there are some limitations, yes. 

Demo: https://jsfiddle.net/q50o1r97/1/

It would be possible with a few modification, though. But for now, I hardwire the forward vector.

Mugen87 committed:
FirstPersonControls: Added .lookAt()

mrdoob commented:
Thanks!

Mugen87 commented:
I think we should update the stackoverflow posts, right?
